proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,556 @@
+
+  [ "......a.b.c"
+  , ""
+  , "......d.e.f"
+  , ".."
+  , "....g.h.i"
+  , "......"
+  ]
+
+The blank line + the whitespace-only line are excluded from the calculation, and we calculate 4 spaces as the shared whitespace prefix, which are removed from every line.
+
+Note that the whitespace preceding the closing ``"""`` is included. This implies that there will be a trailing newline (see the "Trailing newline" example for more information).
+
+Ignore leading characters
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The common prefix calculation ignores all characters preceding the first newline. This means that characters immediately after the ``"""`` delimiter will be included verbatim. The same would occur with a line continuation (since line continuations are collapsed before the prefix calculation).
+
+::
+
+  s =
+    """Line 1
+       Line 2
+    Line 3
+    """
+
+  s_2 =
+    """\
+   \Line 1
+       Line 2
+    Line 3
+    """
+
+  -- equivalent to
+  s' = "Line 1\n   Line 2\nLine 3"
+
+This implies that normal strings could also be written using ``"""``
+
+::
+
+  -- the following are equivalent
+  s = """hello world"""
+  s' = "hello world"
+
+Mixing tabs and spaces
+~~~~~~~~~~~~~~~~~~~~~~
+
+Per the `Haskell report <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-17800010.3>`_, tabs will be treated as 8 spaces in the calculation. If the common prefix is in the middle of a tab (which would only happen if mixing spaces and tabs), the tab will be converted to 8 spaces first. But don't do this.
+
+In the following example, two ``-`` characters will represent 1 tab, and 1 ``.`` character will represent 4 spaces.
+
+::
+
+  s = """
+  ------a
+  ....b
+  .------c
+  ...."""
+
+  -- equivalent to
+  s' = "--a\nb\n.--c\n"
+
+The common whitespace prefix is 16 spaces, so the first line will remove 2 tabs, the second line will remove 16 spaces, and the third line will remove 4 spaces, a tab, then convert the second tab to 8 spaces before removing 4 spaces from it.
+
+Leading newline
+~~~~~~~~~~~~~~~
+
+The specification strips exactly one leading newline, which is the behavior of least surprise for most devs used to multiline strings. To keep the initial newline, add a blank line before the first line:
+
+::
+
+  s =
+    """
+
+    a
+    b
+    c
+    """
+
+  -- equivalent to
+  s' = "\na\nb\nc\n"
+
+Trailing newline
+~~~~~~~~~~~~~~~~
+
+As mentioned in the "Remove common whitespace" example, trailing newlines are naturally included without any explicit rules. As a bonus, it does the same thing that ``unlines`` does. To avoid a trailing newline, put the closing ``"""`` immediately after the last line, or use a line continuation:
+
+::
+
+  x =
+    """
+    a
+    b
+    c"""
+
+  x2 =
+    """
+    a
+    b
+    c\
+    \"""
+
+Indent every line
+~~~~~~~~~~~~~~~~~
+
+To indent every line, use the ``\&`` escape character
+
+::
+
+  s1 =
+    """
+      a
+      b
+      c
+    """
+
+  s1' = "a\nb\nc"
+
+  s2 =
+    """
+    \&  a
```

Comment:
Both are equivalent; I'll clarify this in the doc.

This takes advantage of the `\&` character that's already defined in the report: https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-200002.6. So the behavior described in this section for both `s2` and `s2_2` _comes for free_ based on the existing rules.

Reply:
I understand they're meant to be equivalent, but I just don't understand why you would ever write the version where `\&` only appears on the first line. It makes the string harder to parse by breaking the natural indentation, which is the whole point of this proposal.

Do you have a specific use case where you'd want to write `s2`? If not, I'd strongly suggest to only support `s2_2`. I also would suggest picking a different herald for the indentation, `\&` looks really weird and I expect this indentation-stripping use to be by far the most common case of multi-line strings. (e.g. IntelliJ's Scala extension automatically indents and inserts the leading `|` when you hit enter in a `"""` string)

Reply:
It's not a ”herald for indentation” picked by the PR author, but simply a logical consequence of the rules described further up (note that `\&` is maybe not well known, but it's a normal escape code in Haskell). So these examples will work like this in any case.

If we consider “multi like strings with indentation not fully removed” to be a rarely needed use case, I'd say the proposal is fine: it's _possible_, without making the rules more complicated.

One could argue that `\&` is too obscure and we need something more natural looking. But it would come at the price of having more rules, I fear. Personally, I think the proposed approach is fine - if people learn about Haskell's `\&` along the way, even better.

Reply:
Oh I think I'm beginning to understand. Even though the `\&` escape sequence expands to an empty string, that happens **after** the common prefix computation, which treats it as non-whitespace.

That's rather too subtle I think. For such a core part of the feature I would prefer to have dedicated syntax. Or we could adopt the Java rule that the indentation of the leading/trailing `"""` determines the prefix.

Reply:
I would slightly disagree with the statement that this is a "core part" of the feature. I can't think of a common use case that would require leading whitespace. On the other hand, I can probably think of at least one use case for any character we deem to be the special indentation character. And making the indentation character configurable runs into [other issues](https://github.com/ghc-proposals/ghc-proposals/pull/569#discussion_r1342196233).

IMO This proposal should be reviewed as-is, and we can add support for leading indentation in a later proposal, after we handle the majority of the need for multiline strings.